Source/santabs/SNTBundleService.m

```diff
@@ -122,170 +130,145 @@ - (void)hashBundleBinariesForEvent:(SNTStoredEvent *)event
+  }
 
-  // In the first stage iterate over every file in the tree checking if it is a binary. If so add
-  // it to the fis set for the second stage. Hashing the file while iterating over the filesystem
-  // causes performance issues. Do them separately.
-  while (1) {
+  // Dispatch a block for every file in dirEnum.
+  dispatch_apply(dirEnum.count, self.queue, ^(size_t i) {
     @autoreleasepool {
-      if (breakDir || progress.isCancelled) break;
-
-      // Wait for a processing thread to become available. At this stage we are only reading the
-      // mach_header. If all processing threads are blocking for more than 60 sec bail.
-      if (dispatch_semaphore_wait(sema, dispatch_time(DISPATCH_TIME_NOW, 60 * NSEC_PER_SEC))) {
-        LOGD(@"isExecutable processing threads timeout");
-        return nil;
-      }
+      if (progress.isCancelled) return;
 
-      dispatch_group_async(group,
-                           dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
-        pthread_mutex_lock(&enumeratorMutex);
-        NSString *file = [dirEnum nextObject];
-        fileCount++;
-        pthread_mutex_unlock(&enumeratorMutex);
-
-        if (!file) {
-          breakDir = YES;
-          dispatch_semaphore_signal(sema);
-          return;
-        }
-
-        if ([dirEnum fileAttributes][NSFileType] != NSFileTypeRegular) {
-          dispatch_semaphore_signal(sema);
-          return;
-        }
-
-        NSString *newFile = [event.fileBundlePath stringByAppendingPathComponent:file];
-        SNTFileInfo *fi = [[SNTFileInfo alloc] initWithPath:newFile];
-        if (!fi.isExecutable) {
-          dispatch_semaphore_signal(sema);
-          return;
-        }
-
-        pthread_mutex_lock(&eventsMutex);
-        [fis addObject:fi];
-        binaryCount++;
-        pthread_mutex_unlock(&eventsMutex);
-
-        dispatch_semaphore_signal(sema);
+      dispatch_sync(dispatch_get_main_queue(), ^{
+        p.completedUnitCount++;
       });
-      if (progress && ((fileCount % 500) == 0 || binaryCount > sentBinaryCount)) {
+
+      if (progress && ((i % 500) == 0 || binaryCount > sentBinaryCount)) {
         sentBinaryCount = binaryCount;
         [[self.notifierConnection remoteObjectProxy] updateCountsForEvent:event
                                                               binaryCount:binaryCount
-                                                                fileCount:fileCount];
+                                                                fileCount:i
+                                                              hashedCount:0];
       }
+
+      NSString *file = dirEnum[i];
+      if (!file) return;
+
+      NSString *newFile =
+          [event.fileBundlePath stringByAppendingPathComponent:file].stringByStandardizingPath;
+      SNTFileInfo *fi = [[SNTFileInfo alloc] initWithResolvedPath:newFile error:NULL];
+      if (!fi.isExecutable) return;
+
+      fis[i] = (void *)CFBridgingRetain(fi);
```

Comment:
As this isn't a CF object that you're taking retained ownership of, I think I'd prefer a ```__bridge_retained``` instead. They're identical, it's just that CFBridgingRetain returns a ```CFTypeRef``` while the bridge can be done directly to void:

```objc
fis[i] = (__bridge_retained void *)fi
```